allennlp/modules/similarity_functions/linear.py

```diff
@@ -0,0 +1,131 @@
+    combination done is configurable.
+
+    If the two vectors are ``x`` and ``y``, we allow the following kinds of combinations: ``x``,
+    ``y``, ``x*y``, ``x+y``, ``x-y``, ``x/y``, where each of those binary operations is performed
+    elementwise.  You can list as many combinations as you want, comma separated.  For example, you
+    might give ``x,y,x*y`` as the ``combination`` parameter to this class.  The computed similarity
+    function would then be ``w^T [x; y; x*y] + b``, where ``w`` is a vector of weights, ``b`` is a
+    bias parameter, and ``[;]`` is vector concatenation.
+
+    Note that if you want a bilinear similarity function with a diagonal weight matrix W, where the
+    similarity function is computed as `x * w * y + b` (with `w` the diagonal of `W`), you can
+    accomplish that with this class by using "x*y" for `combination`.
+
+    Parameters
+    ----------
+    tensor_1_dim : ``int``
+        The dimension of the first tensor, ``x``, described above.  This is ``x.size()[-1]`` - the
+        length of the vector that will go into the similarity computation.  We need this so we can
+        build weight vectors correctly.
+    tensor_2_dim : ``int``
+        The dimension of the second tensor, ``y``, described above.  This is ``y.size()[-1]`` - the
+        length of the vector that will go into the similarity computation.  We need this so we can
+        build weight vectors correctly.
+    combination : ``str``, optional (default=``"x,y"``)
+        Described above.
+    activation : ``Callable[[torch.Tensor], torch.Tensor]``, optional (default=``lambda x: x``)
+        An activation function applied after the ``w^T * [x;y] + b`` calculation.  Default is no
+        activation.
+    """
+    def __init__(self,
+                 tensor_1_dim: int,
+                 tensor_2_dim: int,
+                 combination: str = 'x,y',
+                 activation: Callable[[torch.Tensor], torch.Tensor] = lambda x: x) -> None:
+        super(Linear, self).__init__()
+        self._combinations = combination.split(',')
+        combined_dim = self._get_combined_dim(tensor_1_dim, tensor_2_dim)
+        self._weight_vector = Parameter(torch.Tensor(combined_dim))
+        self._bias = Parameter(torch.Tensor(1))
+        self._activation = activation
+
+    @overrides
+    def forward(self, tensor_1: torch.Tensor, tensor_2: torch.Tensor) -> torch.Tensor:
+        combined_tensors = self._combine_tensors(tensor_1, tensor_2)
+
+        # The '@' operator here is torch.matmul, but that's only available in pytorch-0.2.
+        # TODO(mattg): switch to using torch.matmul when a version of pytorch with it is released.
+        # I think it's more clear and less magical.
+        # Also, broadcasting this simple addition is only available in pytorch-0.2.  When that's
+        # ready, change this back to `(dot_product + self._bias)`.
+        dot_product = combined_tensors @ self._weight_vector
```

Comment:
This multiplication (and the corresponding one in `Bilinear` doesn't handle higher-order tensors.  This is causing three of the tests to fail.  I'm still trying to figure out how to do this properly, though if we could upgrade to pytorch 0.2 somehow, I think the issue would just go away...